File: bench/fasta/fasta.go
/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/
 *
 * contributed by The Go Authors.
 * Based on C program by Joern Inge Vestgaarden
 * and Jorge Peixoto de Morais Neto.
 * flag.Arg hack by Isaac Gouy
 */

package main

import (
   "bufio"
   "flag"
   "os"
   "strconv"
)

var out *bufio.Writer

var n = 0

const WIDTH = 60 // Fold lines after WIDTH bytes

func min(a, b int) int {
   if a < b {
      return a
   }
   return b
}

type AminoAcid struct {
   p float64
   c byte
}

func AccumulateProbabilities(genelist []AminoAcid) {
   for i := 1; i < len(genelist); i++ {
      genelist[i].p += genelist[i-1].p
   }
}

// RepeatFasta prints the characters of the byte slice s. When it
// reaches the end of the slice, it goes back to the beginning.
// It stops after generating count characters.
// After each WIDTH characters it prints a newline.
// It assumes that WIDTH <= len(s) + 1.
func RepeatFasta(s []byte, count int) {
   pos := 0
   s2 := make([]byte, len(s)+WIDTH)
   copy(s2, s)
   copy(s2[len(s):], s)
   for count > 0 {
      line := min(WIDTH, count)
      out.Write(s2[pos : pos+line])
      out.WriteByte('\n')
      pos += line
      if pos >= len(s) {
         pos -= len(s)
      }
      count -= line
   }
}

var lastrandom uint32 = 42

const (
   IM = 139968
   IA = 3877
   IC = 29573
)

// Each element of genelist is a struct with a character and
// a floating point number p between 0 and 1.
// RandomFasta generates a random float r and
// finds the first element such that p >= r.
// This is a weighted random selection.
// RandomFasta then prints the character of the array element.
// This sequence is repeated count times.
// Between each WIDTH consecutive characters, the function prints a newline.
func RandomFasta(genelist []AminoAcid, count int) {
   buf := make([]byte, WIDTH+1)
   for count > 0 {
      line := min(WIDTH, count)
      for pos := 0; pos < line; pos++ {
         lastrandom = (lastrandom*IA + IC) % IM
         // Integer to float conversions are faster if the integer is signed.
         r := float64(int(lastrandom)) / IM
         for _, v := range genelist {
            if v.p >= r {
               buf[pos] = v.c
               break
            }
         }
      }
      buf[line] = '\n'
      out.Write(buf[0 : line+1])
      count -= line
   }
}

func main() {
   out = bufio.NewWriter(os.Stdout)
   defer out.Flush()

   flag.Parse()
   if flag.NArg() > 0 { n,_ = strconv.Atoi( flag.Arg(0) ) }

   iub := []AminoAcid{
      AminoAcid{0.27, 'a'},
      AminoAcid{0.12, 'c'},
      AminoAcid{0.12, 'g'},
      AminoAcid{0.27, 't'},
      AminoAcid{0.02, 'B'},
      AminoAcid{0.02, 'D'},
      AminoAcid{0.02, 'H'},
      AminoAcid{0.02, 'K'},
      AminoAcid{0.02, 'M'},
      AminoAcid{0.02, 'N'},
      AminoAcid{0.02, 'R'},
      AminoAcid{0.02, 'S'},
      AminoAcid{0.02, 'V'},
      AminoAcid{0.02, 'W'},
      AminoAcid{0.02, 'Y'},
   }

   homosapiens := []AminoAcid{
      AminoAcid{0.3029549426680, 'a'},
      AminoAcid{0.1979883004921, 'c'},
      AminoAcid{0.1975473066391, 'g'},
      AminoAcid{0.3015094502008, 't'},
   }

   AccumulateProbabilities(iub)
   AccumulateProbabilities(homosapiens)

   alu := []byte(
      "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
         "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
         "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
         "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
         "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
         "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
         "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA")

   out.WriteString(">ONE Homo sapiens alu\n")
   RepeatFasta(alu, 2*n)
   out.WriteString(">TWO IUB ambiguity codes\n")
   RandomFasta(iub, 3*n)
   out.WriteString(">THREE Homo sapiens frequency\n")
   RandomFasta(homosapiens, 5*n)
}

File: bench/fasta/fasta.swift
/* The Computer Language Benchmarks Game
 http://benchmarksgame.alioth.debian.org/
 contributed by Ralph Ganszky
 converted to Swift 3 by Sergo Beruashvili
 */

import Glibc
//import Darwin
typealias AminoAcid = (prob: Double, sym: UInt8)

let IM = 139968
let IA = 3877
let IC = 29573
var seed = 42

let n: Int
if CommandLine.arguments.count > 1 {
    n = Int(CommandLine.arguments[1]) ?? 1000
} else {
    n = 1000
}

let bufferSize = 256*1024
let width = 60

let aluString = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
"AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
var alu = aluString.utf8CString.map({ UInt8($0) })
alu.popLast()

var iub = [
    AminoAcid(0.27, 97), // "a"),
    AminoAcid(0.12, 99), // "c"),
    AminoAcid(0.12, 103), // "g"),
    AminoAcid(0.27, 116), // "t"),
    AminoAcid(0.02, 66), // "B"),
    AminoAcid(0.02, 68), // "D"),
    AminoAcid(0.02, 72), // "H"),
    AminoAcid(0.02, 75), // "K"),
    AminoAcid(0.02, 77), // "M"),
    AminoAcid(0.02, 78), // "N"),
    AminoAcid(0.02, 82), // "R"),
    AminoAcid(0.02, 83), // "S"),
    AminoAcid(0.02, 86), // "V"),
    AminoAcid(0.02, 87), // "W"),
    AminoAcid(0.02, 89), // "Y"),
]

var homosapiens = [
    AminoAcid(0.3029549426680, 97), // "a"),
    AminoAcid(0.1979883004921, 99), // "c"),
    AminoAcid(0.1975473066391, 103), // "g"),
    AminoAcid(0.3015094502008, 116), // "t"),
]

func write(array: [UInt8], ofLen len: Int) {
    var iov = UnsafeMutablePointer<iovec>.allocate(capacity: 1)
    defer {
        iov.deallocate(capacity: 1)
    }
    let data = array.withUnsafeBufferPointer({ p in p })
    iov.pointee.iov_base = UnsafeMutableRawPointer(mutating: data.baseAddress)
    iov.pointee.iov_len = len
    writev(STDOUT_FILENO, iov, 1)
}

func write(buffer: UnsafePointer<Int8>, ofLen len: Int) {
    var iov = UnsafeMutablePointer<iovec>.allocate(capacity: 1)
    defer {
        iov.deallocate(capacity: 1)
    }
    iov.pointee.iov_base = UnsafeMutableRawPointer(mutating: buffer)
    iov.pointee.iov_len = len
    writev(STDOUT_FILENO, iov, 1)
}

func repeatFasta(gene: [UInt8], n: Int) {
    let gene2 = gene + gene
    var buffer = [UInt8](repeating: 10, count: bufferSize)
    var pos = 0
    var rpos = 0
    var cnt = n
    var lwidth = width
    while cnt > 0 {
        if pos + lwidth > buffer.count {
            write(array: buffer, ofLen: pos)
            pos = 0
        }
        if rpos + lwidth > gene.count {
            rpos = rpos % gene.count
        }
        if cnt < lwidth {
            lwidth = cnt
        }
        buffer[pos..<pos+lwidth] = gene2[rpos..<rpos+lwidth]
        buffer[pos+lwidth] = 10
        pos += lwidth + 1
        rpos += lwidth
        cnt -= lwidth
    }
    if pos > 0 && pos < buffer.count {
        buffer[pos] = 10
        write(array: buffer, ofLen: pos)
    } else if pos == buffer.count {
        write(array: buffer, ofLen: pos)
        buffer[0] = 10
        write(array: buffer, ofLen: 1)
    }
}

func search(rnd: Double, within arr: [AminoAcid]) -> UInt8 {
    var low = 0
    var high = arr.count - 1
    while low <= high {
        let mid = low + (high - low) / 2
        if arr[mid].prob >= rnd {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return arr[high+1].sym
}

func accumulateProbabilities( acid: inout [AminoAcid]) {
    for i in 1..<acid.count {
        acid[i].prob += acid[i-1].prob
    }
}

func randomFasta( acid: inout [AminoAcid], _ n: Int) {
    var cnt = n
    accumulateProbabilities(acid: &acid)
    var buffer = [UInt8](repeating: 10, count: bufferSize)
    var pos = 0
    while cnt > 0 {
        var m = cnt
        if m > width {
            m = width
        }
        let f = 1.0 / Double(IM)
        var myrand = seed
        for _ in 0..<m {
            myrand = (myrand * IA + IC) % IM
            let r = Double(myrand) * f
            buffer[pos] = search(rnd: r, within: acid)
            pos += 1
            if pos == buffer.count {
                write(array: buffer, ofLen: pos)
                pos = 0
            }
        }
        seed = myrand
        buffer[pos] = 10
        pos += 1
        if pos == buffer.count {
            write(array: buffer, ofLen: pos)
            pos = 0
        }
        cnt -= m
    }
    if pos > 0 {
        write(array: buffer, ofLen: pos)
    }
}

let one = ">ONE Homo sapiens alu\n"
let oneStr = one.withCString({ s in s })
write(buffer: oneStr, ofLen: one.utf8CString.count - 1)
repeatFasta(gene: alu, n: 2*n)

let two = ">TWO IUB ambiguity codes\n"
let twoStr = two.withCString({ s in s })
write(buffer: twoStr, ofLen: two.utf8CString.count - 1)
randomFasta(acid: &iub,3*n)

let three = ">THREE Homo sapiens frequency\n"
let threeStr = three.withCString({ s in s })
write(buffer: threeStr, ofLen: three.utf8CString.count - 1)
randomFasta(acid: &homosapiens,5*n)
